### ISSUE TITLE
Cannot Run GPU Workloads in Kubeflow Notebooks on RKE2 with NVIDIA Operator

### ISSUE THREAD

**Issue by csanadpoda** [kubeflow/notebooks#522]
### Checks

- [x] I have searched the [existing issues](https://github.com/kubeflow/notebooks/issues).
- [x] My issue is related to one of the components in the [`kubeflow/notebooks`](https://github.com/kubeflow/notebooks) repository.

### Kubeflow Notebooks Version

1.10.0

### Kubeflow Platform

Kubeflow Reference platform installed via the manifests

### Kubernetes Distribution

RKE2 on bare metal

### Kubernetes Version

```shell
Client Version: v1.33.3
Server Version: v1.33.2+rke2r1
Kustomize Version: v5.6.0
```

### Description

# GPU-accelerated notebooks never see GPUs (e.g., `torch.cuda.is_available()` is always False) even though:

- `nvidia.com/gpu` is allocatable
- Node labels and device plugins look fine
- Running a test pod with `runtimeClassName: nvidia` (e.g., CUDA sample) *does* see and use the GPU as expected

## Steps:

- Created Rancher deployment
- Created new Kubeflow cluster
- Used metallb and local-path-provisioner to load balance and set default storage
- Installed the Kubeflow reference platform following the [manifests guide](https://github.com/kubeflow/manifests?tab=readme-ov-file#install-with-a-single-command)
- Created a new VM with a GPU
- Added it to the Cluster
- Deployed the Deploy NVIDIA operator via the [Rancher docs](https://docs.rke2.io/advanced#deploy-nvidia-operator)
    - `nvidia.com/gpu` is allocatable
    - Node labels and device plugins look fine
    - Running a test pod with `runtimeClassName: nvidia` (e.g., CUDA sample) *does* see and use the GPU as expected

The driver installation appears to succeed, and I can see `/run/nvidia/driver/bin/nvidia-smi` present on my GPU node.

- The standard `nvidia-smi` isn’t available in `/usr/bin` unless I manually install from the apt repos, which I’m not supposed to do with GPU Operator.
- Running as `/run/nvidia/driver/bin/nvidia-smi` should work, but it can’t find the necessary libraries.

```
$ sudo find /run/nvidia/driver/ -name ‘libnvidia-ml.so*’
/run/nvidia/driver/usr/lib/x86_64-linux-gnu/libnvidia-ml.so
/run/nvidia/driver/usr/lib/x86_64-linux-gnu/libnvidia-ml.so.1
/run/nvidia/driver/usr/lib/x86_64-linux-gnu/libnvidia-ml.so.570.148.08

$ sudo /run/nvidia/driver/bin/nvidia-smi
NVIDIA-SMI couldn't find libnvidia-ml.so library in your system. Please make sure that the NVIDIA Display Driver is properly installed and present in your system.
Please also try adding directory that contains libnvidia-ml.so to your system PATH.

```
- Opened Kubeflow, was able to create CPU pod and notebook and run Python
- Created GPU pod
- GPU seems to be assigned, but running:

```python
import torch
print("PyTorch version:", torch.__version__)
print("CUDA version (PyTorch sees):", torch.version.cuda)
print("CUDA available:", torch.cuda.is_available())
print("CUDA device count:", torch.cuda.device_count())
print("Device name:", torch.cuda.get_device_name(0) if torch.cuda.is_available() else "No GPU")
```
returns:
```
PyTorch version: 2.5.1+cu124
CUDA version (PyTorch sees): 12.4
CUDA available: False
CUDA device count: 0
Device name: No GPU
```
I’ve read in other [issues](https://github.com/kubeflow/manifests/issues/2683) that one should set default_runtime_name = "nvidia" but that broke my cluster, metallb and a bunch of other services were unable to recover from CrashLoopBackoffs. Resetting to runc solved this.

My Kubeflow deployment should be user facing so I can’t expect people to start new notebooks from YAMLs, it should work just by creating a notebook and requesting a GPU with it. I used the ghcr.io/kubeflow/kubeflow/notebook-servers/jupyter-pytorch-cuda-full:v1.10.0 image which should be sufficient. Plus when someone would be running a CPU pod, using default_runtime_name = "nvidia" would also not necessarily be appropriate.

How can I make it so the GPU gets detected?

Is there a way to somehow inject default_runtime_name = "nvidia" into GPU-supported notebook pods only? What am I missing? Thank you!

### Relevant Logs

```shell
kubectl logs gpu-0 -n kubeflow-user-example-com

/package/admin/s6-overlay/libexec/preinit: info: /run belongs to uid 0 instead of 1000 - fixing it
s6-rc: info: service s6rc-oneshot-runner: starting
s6-rc: info: service s6rc-oneshot-runner successfully started
s6-rc: info: service fix-attrs: starting
s6-rc: info: service fix-attrs successfully started
s6-rc: info: service legacy-cont-init: starting
cont-init: info: running /etc/cont-init.d/01-copy-tmp-home
INFO: Copying contents of '/tmp_home/jovyan' to '/home/jovyan'...
cont-init: info: /etc/cont-init.d/01-copy-tmp-home exited 0
s6-rc: info: service legacy-cont-init successfully started
s6-rc: info: service legacy-services: starting
services-up: info: copying legacy longrun jupyterlab (no readiness notification)
s6-rc: info: service legacy-services successfully started
INFO: starting jupyter...
[W 2025-08-05 17:37:58.393 ServerApp] ServerApp.token config is deprecated in 2.0. Use IdentityProvider.token.
[I 2025-08-05 17:37:58.433 ServerApp] jupyter_lsp | extension was successfully linked.
[I 2025-08-05 17:37:58.437 ServerApp] jupyter_server_mathjax | extension was successfully linked.
[I 2025-08-05 17:37:58.440 ServerApp] jupyter_server_terminals | extension was successfully linked.
[I 2025-08-05 17:37:58.443 ServerApp] jupyterlab | extension was successfully linked.
[I 2025-08-05 17:37:58.443 ServerApp] jupyterlab_git | extension was successfully linked.
[I 2025-08-05 17:37:58.443 ServerApp] nbdime | extension was successfully linked.
[I 2025-08-05 17:37:58.446 ServerApp] notebook | extension was successfully linked.
[I 2025-08-05 17:37:58.447 ServerApp] Writing Jupyter server cookie secret to /tmp/jupyter_runtime/jupyter_cookie_secret
[I 2025-08-05 17:37:58.639 ServerApp] notebook_shim | extension was successfully linked.
[W 2025-08-05 17:37:58.652 ServerApp] All authentication is disabled.  Anyone who can connect to this server will be able to run code.
[I 2025-08-05 17:37:58.652 ServerApp] notebook_shim | extension was successfully loaded.
[I 2025-08-05 17:37:58.654 ServerApp] jupyter_lsp | extension was successfully loaded.
[I 2025-08-05 17:37:58.654 ServerApp] jupyter_server_mathjax | extension was successfully loaded.
[I 2025-08-05 17:37:58.655 ServerApp] jupyter_server_terminals | extension was successfully loaded.
[I 2025-08-05 17:37:58.657 LabApp] JupyterLab extension loaded from /opt/conda/lib/python3.11/site-packages/jupyterlab
[I 2025-08-05 17:37:58.657 LabApp] JupyterLab application directory is /opt/conda/share/jupyter/lab
[I 2025-08-05 17:37:58.658 LabApp] Extension Manager is 'pypi'.
[I 2025-08-05 17:37:58.693 ServerApp] jupyterlab | extension was successfully loaded.
[I 2025-08-05 17:37:58.698 ServerApp] jupyterlab_git | extension was successfully loaded.
[I 2025-08-05 17:37:58.735 ServerApp] nbdime | extension was successfully loaded.
[I 2025-08-05 17:37:58.739 ServerApp] notebook | extension was successfully loaded.
[I 2025-08-05 17:37:58.739 ServerApp] Serving notebooks from local directory: /home/jovyan
[I 2025-08-05 17:37:58.739 ServerApp] Jupyter Server 2.15.0 is running at:
[I 2025-08-05 17:37:58.739 ServerApp] http://gpu-0:8888/notebook/kubeflow-user-example-com/gpu/lab
[I 2025-08-05 17:37:58.739 ServerApp]     http://127.0.0.1:8888/notebook/kubeflow-user-example-com/gpu/lab
[I 2025-08-05 17:37:58.739 ServerApp] Use Control-C to stop this server and shut down all kernels (twice to skip confirmation).
[I 2025-08-05 17:37:58.855 ServerApp] Skipped non-installed server(s): bash-language-server, dockerfile-language-server-nodejs, javascript-typescript-langserver, jedi-language-server, julia-language-server, pyright, python-language-server, python-lsp-server, r-languageserver, sql-language-server, texlab, typescript-language-server, unified-language-server, vscode-css-languageserver-bin, vscode-html-languageserver-bin, vscode-json-languageserver-bin, yaml-language-server
[I 2025-08-05 17:38:40.123 ServerApp] 302 GET /notebook/kubeflow-user-example-com/gpu/ (@127.0.0.6) 1.23ms
[I 2025-08-05 17:38:42.578 LabApp] Build is up to date
[I 2025-08-05 17:38:42.813 ServerApp] Kernel started: 534ba690-0f44-43fc-9313-9dfc81a72f37
[I 2025-08-05 17:38:43.402 ServerApp] Connecting to kernel 534ba690-0f44-43fc-9313-9dfc81a72f37.
[I 2025-08-05 17:38:43.505 ServerApp] Connecting to kernel 534ba690-0f44-43fc-9313-9dfc81a72f37.
[I 2025-08-05 17:38:43.673 ServerApp] Connecting to kernel 534ba690-0f44-43fc-9313-9dfc81a72f37.
[I 2025-08-05 17:40:43.274 ServerApp] Saving file at /Untitled.ipynb


kubectl logs nbody-gpu-benchmark -n default
Run "nbody -benchmark [-numbodies=<numBodies>]" to measure performance.
	-fullscreen       (run n-body simulation in fullscreen mode)
	-fp64             (use double precision floating point values for simulation)
	-hostmem          (stores simulation data in host memory)
	-benchmark        (run benchmark to measure performance) 
	-numbodies=<N>    (number of bodies (>= 1) to run in simulation) 
	-device=<d>       (where d=0,1,2.... for the CUDA device to use)
	-numdevices=<i>   (where i=(number of CUDA devices > 0) to use for simulation)
	-compare          (compares simulation results running once on the default GPU and once on the CPU)
	-cpu              (run n-body simulation on the CPU)
	-tipsy=<file.bin> (load a tipsy model file for simulation)

NOTE: The CUDA Samples are not meant for performance measurements. Results may vary when GPU Boost is enabled.

> Windowed mode
> Simulation data stored in video memory
> Single precision floating point simulation
> 1 Devices used for simulation
GPU Device 0: "Turing" with compute capability 7.5

> Compute 7.5 CUDA device: [Tesla T4]
40960 bodies, total time for 10 iterations: 96.007 ms
= 174.750 billion interactions per second
= 3494.998 single-precision GFLOP/s at 20 flops per interaction
```

**Comment by giosdas** [comment]
Hello,
I'm facing the same issue.
@csanadpoda  did you find a workaround?

Thanks
Regards

**Comment by csanadpoda** [comment]
> Hello, I'm facing the same issue. [@csanadpoda](https://github.com/csanadpoda) did you find a workaround?
> 
> Thanks Regards

Yes, I installed Kyverno via theri helm chart and added this:

```
apiVersion: kyverno.io/v1
kind: ClusterPolicy
metadata:
  name: add-nvidia-runtime-for-gpu-pods
spec:
  rules:
    - name: add-runtimeclass-to-gpu-pods
      match:
        any:
          - resources:
              kinds:
                - Pod
              namespaces:
                - kubeflow-user-example-com
      preconditions:
        all:
          - key: "{{ request.operation }}"
            operator: In
            value: ["CREATE","UPDATE"]
          - key: "{{ request.object.spec.runtimeClassName || '' }}"
            operator: Equals
            value: ""
          - key: '{{ request.object.spec.containers[? resources.limits."nvidia.com/gpu" != null || resources.requests."nvidia.com/gpu" != null] | length(@) }}'
            operator: GreaterThan
            value: 0
      mutate:
        patchStrategicMerge:
          spec:
            runtimeClassName: nvidia
```
It will inject the nvidia runtime for notebook pods that request an nvidia gpu.

**Comment by giosdas** [comment]
@csanadpoda Thank you very much!!! It works!
